webrtc/RTCTransport-helper.js

```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+/*
+  Code using this helper should also include RTCPeerConnection-helper.js
+  in the main HTML file
+
+  The following helper functions are called from RTCPeerConnection-helper.js:
+    getTrackFromUserMedia
+    exchangeIceCandidates
+    doSignalingHandshake
+*/
+
+// Wait for all DTLS and ICE transport of a peer connection
+// to turn to connected or completed state. When the RTCPeerConnectionState
+// is "connected", all RTCIceTransports and RTCDtlsTransports are in the
+// "connected", "completed" or "closed" state and at least one of them is in the
+// "connected" or "completed" state.
+async function waitForConnectedState(pc) {
+  if (pc.connectionState === 'connected') {
+    return;
+  }
+
+  return new Promise((resolve, reject) => {
+    pc.addEventListener('connectionstatechange', () => {
+      const { connectionState } = pc;
+
+      if (connectionState === 'connected') {
+        resolve();
+      } else if (['closed', 'failed'].includes(connectionState)) {
+        reject(new Error(`one of DTLS/ICE transports transition to unexpected state ${connectionState}`));
```

Comment:
I advocate [wrapping non-promise code as tightly as possible](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#Creating_a_Promise_around_an_old_callback_API), which means avoiding any kind of application code inside promise constructors.

The WPT test harness adds a second reason: the awkward `t.step_func()` wrapper is needed to correlate spurious event handlers with a particular test, and it's missing here. Unfortunately, this would mean adding `t` as an additional argument to `waitForConnectedState`, which gets ugly.

Instead, I would recommend something like this:
```js
while (pc.connectionState != 'connected') {
  await new Promise(r => pc.addEventListener('connectionstatechange', r, {once: true}));
  if (['closed', 'failed'].includes(pc.connectionState)) {
    throw new Error(`DTLS/ICE transition to unexpected state ${pc.connectionState}`);
  }
}
```

Reply:
I agree with your proposition that inside the promise constructor the code should be as minimal as possible. However I do not agree with the proposed solution. Given that the timing for promise resolution is non-deterministic, you may miss an event firing if after the first event, a second event fires before the continuation following the await is executed. Instead of keeping the promise constructor absolute minimal, I think the better way to interpret is that code inside the promise constructor should be more carefully scrutinized, and having less code is a way for verifying the correctness.

I have added a try catch inside the callback to make sure the promise does not get unresolved forever due to exceptions. There are definitely better ways to do this, but given the many conflicting constraints we have for writing tests, I think that should be enough for this particular case.